src/storage/createInMemoryStorage.ts

```diff
@@ -1,72 +1,26 @@
-import { Query } from '../query/createQuery';
-import { StoreOperation, CrudOptions, StoreOptions } from '../store/createStore';
-import compose, { ComposeFactory } from 'dojo-compose/compose';
+import compose from 'dojo-compose/compose';
 import Promise from 'dojo-shim/Promise';
 import WeakMap from 'dojo-shim/WeakMap';
 import Map from 'dojo-shim/Map';
-import Patch from '../patch/Patch';
-import { PatchMapEntry } from '../patch/Patch';
 import { duplicate } from 'dojo-core/lang';
-
-export interface Storage<T, O extends CrudOptions> {
-	identify(items: T[]|T): string[];
-	createId(): Promise<string>;
-	fetch(): Promise<T[]>;
-	fetch<V>(query: Query<T, V>): Promise<V[]>;
-	get(ids: string[]): Promise<T[]>;
-	put(items: T[], options?: O): Promise<UpdateResults<T>>;
-	add(items: T[], options?: O): Promise<UpdateResults<T>>;
-	delete(ids: string[]): Promise<UpdateResults<T>>;
-	patch(updates: { id: string; patch: Patch<T, T> }[], options?: O): Promise<UpdateResults<T>>;
-	isUpdate(item: T): Promise<{ isUpdate: boolean; item: T, id: string }>;
-}
+import Patch from '../patch/Patch';
+import { StoreOperation, CrudOptions, StoreOptions } from '../store/createStore';
+import { Query } from '../query/createQuery';
+import createBaseStorage, { Storage, StorageFactory, UpdateResults } from './createBaseStorage';
 
 export interface InMemoryStorageState<T> {
-	idProperty?: string;
-	idFunction?: (item: T) => string;
 	nextId?: number;
 	data?: T[];
 	index: Map<string, number>;
 }
 
-export type CrudArgument<T> = T | string | PatchMapEntry<T, T>;
-export interface UpdateResults<T> {
-	currentItems?: T[];
-	failedData?: CrudArgument<T>[];
-	successfulData: T[] | string[];
-	type: StoreOperation;
-}
-
 const instanceStateMap = new WeakMap<Storage<{}, {}>, InMemoryStorageState<{}>>();
 
-export interface StorageFactory extends ComposeFactory<Storage<{}, {}>, StoreOptions<{}, CrudOptions>> {
-	<T extends {}, O extends CrudOptions>(options?: O): Storage<T, O>;
-}
-
 export interface InMemoryStorageFactory extends StorageFactory {
 	<T>(options?: StoreOptions<T, CrudOptions>): Storage<T, CrudOptions>;
 }
 
-type IdObject = { [ index: string ]: string; id: string };
-
-const createInMemoryStorage: InMemoryStorageFactory = compose<Storage<IdObject, CrudOptions>, StoreOptions<{}, CrudOptions>>({
-	identify(this: Storage<{}, {}>, items: IdObject[]| IdObject): string[] {
-		const state = instanceStateMap.get(this);
-		const itemArray = Array.isArray(items) ? <IdObject []> items : [ <IdObject> items ];
-		if (state.idProperty) {
-			return itemArray.map((item) => {
-				return item[state.idProperty];
-			});
-		}
-		else if (state.idFunction) {
-			return itemArray.map(state.idFunction);
-		}
-		else {
-			return itemArray.map(function(item) {
-				return item.id;
-			});
-		}
-	},
+const createInMemoryStorage: InMemoryStorageFactory = createBaseStorage.mixin(compose({
```

Comment:
Yeah. That's a good idea and I tried it. But it doesn't seem to work. You can find the change in [this branch](https://github.com/lzhoucs/stores/blob/rest-storage-feedback/src/storage/createRestStorage.ts#L76)